File: backend/app/routers/open_field.py
"""Router FastAPI — Croisement Open Field."""
from datetime import date
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session, selectinload

from app.database import get_db
from app.models.all_models import (
    ProjetOpenField, PlanteMereOpenField, PlantePereOpenField,
    Variete, Pollen, Plant, Culture, PackGraine, Breeder, Graine,
)
from app.schemas.open_field import (
    ProjetOpenFieldCreate, ProjetOpenFieldUpdate, ProjetOpenFieldRead,
    PlanteMereCreate, PlanteMereUpdate, PlanteMereRead,
    PlantePereCreate, PlantePereUpdate, PlantePereRead,
    RecolteInput,
)

router = APIRouter(prefix="/api/open-field", tags=["open_field"])


# ─── Helpers ──────────────────────────────────────────────────────────────────

def _enrich_mere(m: PlanteMereOpenField, peres_map: dict = None) -> dict:
    d = {c.name: getattr(m, c.name) for c in m.__table__.columns}
    d["variete_nom"]  = m.variete.nom_variete if m.variete else None
    d["pollen_nom"]   = m.pollen.nom_pollen   if m.pollen  else None
    d["plant_label"]  = f"Plant #{m.plant.numero_plant}" if m.plant else None
    if peres_map and m.id_peres:
        d["peres_labels"] = [peres_map.get(pid, f"#{pid}") for pid in (m.id_peres or [])]
    else:
        d["peres_labels"] = []
    return d

def _enrich_projet(p: ProjetOpenField) -> dict:
    d = {c.name: getattr(p, c.name) for c in p.__table__.columns}
    d["culture_nom"]      = p.culture.nom if p.culture else None
    peres_map = {
        pe.id_pere: (pe.variete.nom_variete if pe.variete else pe.nom_libre or f"Mâle #{pe.id_pere}")
        for pe in p.peres
    }
    d["peres"]            = [{"id_pere": pe.id_pere, "id_projet": pe.id_projet,
                               "id_variete": pe.id_variete, "nom_libre": pe.nom_libre,
                               "notes": pe.notes, "created_at": pe.created_at,
                               "variete_nom": pe.variete.nom_variete if pe.variete else None}
                              for pe in p.peres]
    d["meres"]            = [_enrich_mere(m, peres_map) for m in p.meres]
    d["nb_meres"]         = len(p.meres)
    d["nb_peres"]         = len(p.peres)
    d["nb_graines_total"] = sum(m.nb_graines or 0 for m in p.meres)
    return d

def _load_projet(db: Session, projet_id: int) -> ProjetOpenField:
    p = (
        db.query(ProjetOpenField)
        .options(
            selectinload(ProjetOpenField.meres).selectinload(PlanteMereOpenField.variete),
            selectinload(ProjetOpenField.meres).selectinload(PlanteMereOpenField.pollen),
            selectinload(ProjetOpenField.meres).selectinload(PlanteMereOpenField.plant),
            selectinload(ProjetOpenField.meres).selectinload(PlanteMereOpenField.packgraine),
            selectinload(ProjetOpenField.peres).selectinload(PlantePereOpenField.variete),
            selectinload(ProjetOpenField.culture),
        )
        .filter(ProjetOpenField.id_projet == projet_id)
        .first()
    )
    if not p:
        raise HTTPException(status_code=404, detail="Projet open field non trouvé")
    return p


# ─── Projets CRUD ─────────────────────────────────────────────────────────────

@router.get("", response_model=list[ProjetOpenFieldRead])
def list_projets(db: Session = Depends(get_db), statut: Optional[str] = None):
    q = (
        db.query(ProjetOpenField)
        .options(
            selectinload(ProjetOpenField.meres).selectinload(PlanteMereOpenField.variete),
            selectinload(ProjetOpenField.meres).selectinload(PlanteMereOpenField.pollen),
            selectinload(ProjetOpenField.meres).selectinload(PlanteMereOpenField.plant),
            selectinload(ProjetOpenField.meres).selectinload(PlanteMereOpenField.packgraine),
            selectinload(ProjetOpenField.peres).selectinload(PlantePereOpenField.variete),
            selectinload(ProjetOpenField.culture),
        )
        .order_by(ProjetOpenField.created_at.desc())
    )
    if statut:
        q = q.filter(ProjetOpenField.statut == statut)
    return [_enrich_projet(p) for p in q.all()]


@router.get("/{projet_id}", response_model=ProjetOpenFieldRead)
def get_projet(projet_id: int, db: Session = Depends(get_db)):
    return _enrich_projet(_load_projet(db, projet_id))


@router.post("", response_model=ProjetOpenFieldRead, status_code=201)
def create_projet(data: ProjetOpenFieldCreate, db: Session = Depends(get_db)):
    projet = ProjetOpenField(**data.model_dump())
    db.add(projet)
    db.commit()
    db.refresh(projet)
    return _enrich_projet(_load_projet(db, projet.id_projet))


@router.patch("/{projet_id}", response_model=ProjetOpenFieldRead)
def update_projet(projet_id: int, data: ProjetOpenFieldUpdate, db: Session = Depends(get_db)):
    projet = _load_projet(db, projet_id)
    for k, v in data.model_dump(exclude_unset=True).items():
        setattr(projet, k, v)
    db.commit()
    return _enrich_projet(_load_projet(db, projet_id))


@router.delete("/{projet_id}", status_code=204)
def delete_projet(projet_id: int, db: Session = Depends(get_db)):
    projet = _load_projet(db, projet_id)
    db.delete(projet)
    db.commit()


# ─── Mères CRUD ───────────────────────────────────────────────────────────────

@router.post("/{projet_id}/meres", response_model=ProjetOpenFieldRead, status_code=201)
def add_mere(projet_id: int, data: PlanteMereCreate, db: Session = Depends(get_db)):
    _load_projet(db, projet_id)  # vérif existence
    mere = PlanteMereOpenField(id_projet=projet_id, **data.model_dump())
    db.add(mere)
    db.commit()
    return _enrich_projet(_load_projet(db, projet_id))


@router.patch("/{projet_id}/meres/{mere_id}", response_model=ProjetOpenFieldRead)
def update_mere(projet_id: int, mere_id: int, data: PlanteMereUpdate, db: Session = Depends(get_db)):
    mere = db.query(PlanteMereOpenField).filter(
        PlanteMereOpenField.id_mere == mere_id,
        PlanteMereOpenField.id_projet == projet_id,
    ).first()
    if not mere:
        raise HTTPException(status_code=404, detail="Mère non trouvée")
    for k, v in data.model_dump(exclude_unset=True).items():
        setattr(mere, k, v)
    db.commit()
    return _enrich_projet(_load_projet(db, projet_id))


@router.delete("/{projet_id}/meres/{mere_id}", status_code=204)
def delete_mere(projet_id: int, mere_id: int, db: Session = Depends(get_db)):
    mere = db.query(PlanteMereOpenField).filter(
        PlanteMereOpenField.id_mere == mere_id,
        PlanteMereOpenField.id_projet == projet_id,
    ).first()
    if not mere:
        raise HTTPException(status_code=404, detail="Mère non trouvée")
    db.delete(mere)
    db.commit()


# ─── Récolte ──────────────────────────────────────────────────────────────────

@router.post("/{projet_id}/meres/{mere_id}/recolte", response_model=ProjetOpenFieldRead)
def recolte_mere(projet_id: int, mere_id: int, data: RecolteInput, db: Session = Depends(get_db)):
    mere = db.query(PlanteMereOpenField).options(
        selectinload(PlanteMereOpenField.variete),
        selectinload(PlanteMereOpenField.pollen),
    ).filter(
        PlanteMereOpenField.id_mere == mere_id,
        PlanteMereOpenField.id_projet == projet_id,
    ).first()
    if not mere:
        raise HTTPException(status_code=404, detail="Mère non trouvée")

    mere.date_recolte    = data.date_recolte or date.today()
    mere.nb_graines      = data.nb_graines
    mere.poids_graines_g = data.poids_graines_g
    mere.qualite_graines = data.qualite_graines

    # Créer un PackGraine + nouvelle Variete + Graines individuelles si demandé
    if data.creer_pack and data.nb_graines:
        projet = db.query(ProjetOpenField).filter(ProjetOpenField.id_projet == projet_id).first()

        # ── Composer le nom du croisement ──
        nom_mere = (mere.variete.nom_variete if mere.variete else None) or mere.nom_phenotype or "Inconnue"
        peres_noms = []
        if mere.id_peres:
            peres_db = db.query(PlantePereOpenField).filter(
                PlantePereOpenField.id_projet == projet_id,
                PlantePereOpenField.id_pere.in_(mere.id_peres)
            ).all()
            for pe in peres_db:
                peres_noms.append(pe.variete.nom_variete if pe.variete else pe.nom_libre or f"Mâle #{pe.id_pere}")
        if not peres_noms and mere.nom_pere_libre:
            peres_noms.append(mere.nom_pere_libre)
        if not peres_noms:
            peres_noms.append("inconnu")

        saison = projet.saison or str(date.today().year)
        nom_croisement = data.nom_variete_croisement or f"{nom_mere} × {' / '.join(peres_noms)} (OF {saison})"
        croisement_desc = f"♀ {nom_mere}  ×  ♂ {', '.join(peres_noms)} — Open Field {saison}"

        # ── Créer la nouvelle Variete ──
        nouvelle_variete = Variete(
            nom_variete=nom_croisement,
            croisement_variete=croisement_desc,
            informations_variete=f"Issu du projet open field : {projet.nom}",
        )
        db.add(nouvelle_variete)
        db.flush()

        # ── PackGraine ──
        pack = PackGraine(
            id_fournisseur=None,
            nbr_graines=data.nb_graines,
            prix_achat=0,
            date_achat=data.date_recolte or date.today(),
        )
        db.add(pack)
        db.flush()
        mere.id_packgraine = pack.id_packgraine

        # ── Graines individuelles liées à la nouvelle variété ──
        for _ in range(data.nb_graines):
            g = Graine(
                id_breeder=None,
                id_variete=nouvelle_variete.id_variete,
                id_packgraine=pack.id_packgraine,
                types_graines="Régulière",
                date_achat=data.date_recolte or date.today(),
                prix_achat=0,
                utilisee=False,
            )
            db.add(g)

    db.commit()

    # Passer le projet en "recolte" si toutes les mères ont une récolte
    projet = _load_projet(db, projet_id)
    if all(m.date_recolte for m in projet.meres):
        projet.statut = "recolte"
        db.commit()

    return _enrich_projet(_load_projet(db, projet_id))

# ─── Pères CRUD ───────────────────────────────────────────────────────────────

@router.post("/{projet_id}/peres", response_model=ProjetOpenFieldRead, status_code=201)
def add_pere(projet_id: int, data: PlantePereCreate, db: Session = Depends(get_db)):
    _load_projet(db, projet_id)
    pere = PlantePereOpenField(id_projet=projet_id, **data.model_dump())
    db.add(pere)
    db.commit()
    return _enrich_projet(_load_projet(db, projet_id))


@router.patch("/{projet_id}/peres/{pere_id}", response_model=ProjetOpenFieldRead)
def update_pere(projet_id: int, pere_id: int, data: PlantePereUpdate, db: Session = Depends(get_db)):
    pere = db.query(PlantePereOpenField).filter(
        PlantePereOpenField.id_pere == pere_id,
        PlantePereOpenField.id_projet == projet_id,
    ).first()
    if not pere:
        raise HTTPException(status_code=404, detail="Pere non trouve")
    for k, v in data.model_dump(exclude_unset=True).items():
        setattr(pere, k, v)
    db.commit()
    return _enrich_projet(_load_projet(db, projet_id))


@router.delete("/{projet_id}/peres/{pere_id}", status_code=204)
def delete_pere(projet_id: int, pere_id: int, db: Session = Depends(get_db)):
    pere = db.query(PlantePereOpenField).filter(
        PlantePereOpenField.id_pere == pere_id,
        PlantePereOpenField.id_projet == projet_id,
    ).first()
    if not pere:
        raise HTTPException(status_code=404, detail="Pere non trouve")
    db.delete(pere)
    db.commit()

File: frontend/src/local/handlers/open-field.ts
// ─── Handlers locaux : Open Field (projets, mères, pères, récolte) ────────────
// Miroir de backend/app/routers/open_field.py
import { route } from '../router'
import { query, one, oneOr404, insert, updateById, run, jsonify } from '../helpers'
import { Row, todayISO } from './cultures-helpers'

const num = (v: unknown): number | null => (v === null || v === undefined ? null : Number(v))

const PROJET_FIELDS = ['nom', 'saison', 'lieu', 'conditions', 'id_culture', 'statut', 'description', 'notes']
const MERE_FIELDS = ['id_variete', 'nom_phenotype', 'id_plant', 'date_pollinisation', 'methode_pollinisation',
  'pere_identifie', 'id_pollen', 'nom_pere_libre', 'id_peres', 'notes_pollinisation', 'date_recolte',
  'nb_graines', 'poids_graines_g', 'qualite_graines', 'id_packgraine', 'notes']
const PERE_FIELDS = ['id_variete', 'nom_libre', 'notes']

async function enrichProjet(p: Row): Promise<Row> {
  let cultureNom: unknown = null
  if (p.id_culture) {
    const c = await one<Row>('SELECT nom FROM "Culture" WHERE id_culture = ?', [p.id_culture])
    cultureNom = c?.nom ?? null
  }
  const peres = await query<Row>('SELECT * FROM "PlantePereOpenField" WHERE id_projet = ?', [p.id_projet])
  const peresOut: Row[] = []
  const peresMap = new Map<number, string>()
  for (const pe of peres) {
    let varieteNom: unknown = null
    if (pe.id_variete) {
      const v = await one<Row>('SELECT nom_variete FROM "Variete" WHERE id_variete = ?', [pe.id_variete])
      varieteNom = v?.nom_variete ?? null
    }
    peresMap.set(Number(pe.id_pere), String(varieteNom ?? pe.nom_libre ?? `Mâle #${pe.id_pere}`))
    peresOut.push({
      id_pere: pe.id_pere, id_projet: pe.id_projet, id_variete: pe.id_variete,
      nom_libre: pe.nom_libre, notes: pe.notes, created_at: pe.created_at,
      variete_nom: varieteNom,
    })
  }
  const meres = await query<Row>('SELECT * FROM "PlanteMereOpenField" WHERE id_projet = ?', [p.id_projet])
  const meresOut: Row[] = []
  for (const m of meres) {
    jsonify(m, ['id_peres'])
    let varieteNom: unknown = null
    if (m.id_variete) {
      const v = await one<Row>('SELECT nom_variete FROM "Variete" WHERE id_variete = ?', [m.id_variete])
      varieteNom = v?.nom_variete ?? null
    }
    let pollenNom: unknown = null
    if (m.id_pollen) {
      const po = await one<Row>('SELECT nom_pollen FROM "Pollen" WHERE id_pollen = ?', [m.id_pollen])
      pollenNom = po?.nom_pollen ?? null
    }
    let plantLabel: unknown = null
    if (m.id_plant) {
      const pl = await one<Row>('SELECT numero_plant FROM "Plant" WHERE id_plant = ?', [m.id_plant])
      if (pl) plantLabel = `Plant #${pl.numero_plant}`
    }
    const idPeres = (m.id_peres as number[] | null) ?? []
    meresOut.push({
      ...m,
      pere_identifie: !!Number(m.pere_identifie ?? 0),
      poids_graines_g: num(m.poids_graines_g),
      variete_nom: varieteNom,
      pollen_nom: pollenNom,
      plant_label: plantLabel,
      peres_labels: idPeres.map(pid => peresMap.get(Number(pid)) ?? `#${pid}`),
    })
  }
  return {
    ...p,
    culture_nom: cultureNom,
    peres: peresOut,
    meres: meresOut,
    nb_meres: meresOut.length,
    nb_peres: peresOut.length,
    nb_graines_total: meres.reduce((s, m) => s + Number(m.nb_graines ?? 0), 0),
  }
}

async function loadProjet(id: unknown): Promise<Row> {
  return oneOr404('SELECT * FROM "ProjetOpenField" WHERE id_projet = ?', [id], 'Projet open field non trouvé')
}

// ── Projets ───────────────────────────────────────────────────────────────────

route('GET', '/open-field', async ({ query: q }) => {
  const statut = q.get('statut')
  const projets = statut
    ? await query<Row>('SELECT * FROM "ProjetOpenField" WHERE statut = ? ORDER BY created_at DESC', [statut])
    : await query<Row>('SELECT * FROM "ProjetOpenField" ORDER BY created_at DESC')
  const out: Row[] = []
  for (const p of projets) out.push(await enrichProjet(p))
  return { data: out }
})

route('GET', '/open-field/:id', async ({ params }) => ({
  data: await enrichProjet(await loadProjet(params.id)),
}))

route('POST', '/open-field', async ({ body }) => {
  const p = body as Row
  const obj: Row = { created_at: new Date().toISOString() }
  for (const f of PROJET_FIELDS) obj[f] = p[f] ?? null
  const id = await insert('ProjetOpenField', obj)
  return { status: 201, data: await enrichProjet((await one<Row>('SELECT * FROM "ProjetOpenField" WHERE id_projet = ?', [id]))!) }
})

route('PATCH', '/open-field/:id', async ({ params, body }) => {
  await loadProjet(params.id)
  const p = body as Row
  const upd: Row = {}
  for (const f of PROJET_FIELDS) if (f in p) upd[f] = p[f]
  await updateById('ProjetOpenField', 'id_projet', params.id, upd)
  return { data: await enrichProjet((await one<Row>('SELECT * FROM "ProjetOpenField" WHERE id_projet = ?', [params.id]))!) }
})

route('DELETE', '/open-field/:id', async ({ params }) => {
  await loadProjet(params.id)
  await run('DELETE FROM "PlanteMereOpenField" WHERE id_projet = ?', [params.id])
  await run('DELETE FROM "PlantePereOpenField" WHERE id_projet = ?', [params.id])
  await run('DELETE FROM "ProjetOpenField" WHERE id_projet = ?', [params.id])
  return { status: 204 }
})

// ── Mères ─────────────────────────────────────────────────────────────────────

route('POST', '/open-field/:id/meres', async ({ params, body }) => {
  await loadProjet(params.id)
  const p = body as Row
  const obj: Row = { id_projet: Number(params.id), created_at: new Date().toISOString() }
  for (const f of MERE_FIELDS) obj[f] = p[f] ?? null
  await insert('PlanteMereOpenField', obj)
  return { status: 201, data: await enrichProjet((await one<Row>('SELECT * FROM "ProjetOpenField" WHERE id_projet = ?', [params.id]))!) }
})

route('PATCH', '/open-field/:projet_id/meres/:mere_id', async ({ params, body }) => {
  await oneOr404('SELECT * FROM "PlanteMereOpenField" WHERE id_mere = ? AND id_projet = ?',
    [params.mere_id, params.projet_id], 'Mère non trouvée')
  const p = body as Row
  const upd: Row = {}
  for (const f of MERE_FIELDS) if (f in p) upd[f] = p[f]
  await updateById('PlanteMereOpenField', 'id_mere', params.mere_id, upd)
  return { data: await enrichProjet((await one<Row>('SELECT * FROM "ProjetOpenField" WHERE id_projet = ?', [params.projet_id]))!) }
})

route('DELETE', '/open-field/:projet_id/meres/:mere_id', async ({ params }) => {
  await oneOr404('SELECT * FROM "PlanteMereOpenField" WHERE id_mere = ? AND id_projet = ?',
    [params.mere_id, params.projet_id], 'Mère non trouvée')
  await run('DELETE FROM "PlanteMereOpenField" WHERE id_mere = ?', [params.mere_id])
  return { status: 204 }
})

// ── Récolte d'une mère ────────────────────────────────────────────────────────

route('POST', '/open-field/:projet_id/meres/:mere_id/recolte', async ({ params, body }) => {
  const mere = await oneOr404<Row>('SELECT * FROM "PlanteMereOpenField" WHERE id_mere = ? AND id_projet = ?',
    [params.mere_id, params.projet_id], 'Mère non trouvée')
  jsonify(mere, ['id_peres'])
  const p = body as Row
  const dateRecolte = p.date_recolte ? String(p.date_recolte) : todayISO()

  await updateById('PlanteMereOpenField', 'id_mere', params.mere_id, {
    date_recolte: dateRecolte,
    nb_graines: p.nb_graines ?? null,
    poids_graines_g: p.poids_graines_g ?? null,
    qualite_graines: p.qualite_graines ?? null,
  })

  if (p.creer_pack && p.nb_graines) {
    const projet = await one<Row>('SELECT * FROM "ProjetOpenField" WHERE id_projet = ?', [params.projet_id])

    // Nom du croisement
    let nomMere = mere.nom_phenotype ? String(mere.nom_phenotype) : 'Inconnue'
    if (mere.id_variete) {
      const v = await one<Row>('SELECT nom_variete FROM "Variete" WHERE id_variete = ?', [mere.id_variete])
      if (v?.nom_variete) nomMere = String(v.nom_variete)
    }
    const peresNoms: string[] = []
    const idPeres = (mere.id_peres as number[] | null) ?? []
    if (idPeres.length) {
      const peres = await query<Row>(
        `SELECT pe.*, v.nom_variete FROM "PlantePereOpenField" pe
         LEFT JOIN "Variete" v ON v.id_variete = pe.id_variete
         WHERE pe.id_projet = ? AND pe.id_pere IN (${idPeres.map(() => '?').join(',')})`,
        [params.projet_id, ...idPeres])
      for (const pe of peres) {
        peresNoms.push(String(pe.nom_variete ?? pe.nom_libre ?? `Mâle #${pe.id_pere}`))
      }
    }
    if (!peresNoms.length && mere.nom_pere_libre) peresNoms.push(String(mere.nom_pere_libre))
    if (!peresNoms.length) peresNoms.push('inconnu')

    const saison = projet?.saison ? String(projet.saison) : String(new Date().getFullYear())
    const nomCroisement = p.nom_variete_croisement
      ? String(p.nom_variete_croisement)
      : `${nomMere} × ${peresNoms.join(' / ')} (OF ${saison})`
    const croisementDesc = `♀ ${nomMere}  ×  ♂ ${peresNoms.join(', ')} — Open Field ${saison}`

    const idVariete = await insert('Variete', {
      nom_variete: nomCroisement,
      croisement_variete: croisementDesc,
      informations_variete: `Issu du projet open field : ${projet?.nom ?? ''}`,
    })
    const idPack = await insert('PackGraine', {
      id_fournisseur: null,
      nbr_graines: p.nb_graines,
      prix_achat: 0,
      date_achat: dateRecolte,
    })
    await run('UPDATE "PlanteMereOpenField" SET id_packgraine = ? WHERE id_mere = ?', [idPack, params.mere_id])

    for (let i = 0; i < Number(p.nb_graines); i++) {
      await insert('Graine', {
        id_breeder: null,
        id_variete: idVariete,
        id_packgraine: idPack,
        types_graines: 'Régulière',
        date_achat: dateRecolte,
        prix_achat: 0,
        utilisee: false,
      })
    }
  }

  // Projet → "recolte" si toutes les mères sont récoltées
  const restantes = await query<Row>(
    'SELECT id_mere FROM "PlanteMereOpenField" WHERE id_projet = ? AND date_recolte IS NULL', [params.projet_id])
  if (!restantes.length) {
    await run('UPDATE "ProjetOpenField" SET statut = ? WHERE id_projet = ?', ['recolte', params.projet_id])
  }

  return { data: await enrichProjet((await one<Row>('SELECT * FROM "ProjetOpenField" WHERE id_projet = ?', [params.projet_id]))!) }
})

// ── Pères ─────────────────────────────────────────────────────────────────────

route('POST', '/open-field/:id/peres', async ({ params, body }) => {
  await loadProjet(params.id)
  const p = body as Row
  const obj: Row = { id_projet: Number(params.id), created_at: new Date().toISOString() }
  for (const f of PERE_FIELDS) obj[f] = p[f] ?? null
  await insert('PlantePereOpenField', obj)
  return { status: 201, data: await enrichProjet((await one<Row>('SELECT * FROM "ProjetOpenField" WHERE id_projet = ?', [params.id]))!) }
})

route('PATCH', '/open-field/:projet_id/peres/:pere_id', async ({ params, body }) => {
  await oneOr404('SELECT * FROM "PlantePereOpenField" WHERE id_pere = ? AND id_projet = ?',
    [params.pere_id, params.projet_id], 'Pere non trouve')
  const p = body as Row
  const upd: Row = {}
  for (const f of PERE_FIELDS) if (f in p) upd[f] = p[f]
  await updateById('PlantePereOpenField', 'id_pere', params.pere_id, upd)
  return { data: await enrichProjet((await one<Row>('SELECT * FROM "ProjetOpenField" WHERE id_projet = ?', [params.projet_id]))!) }
})

route('DELETE', '/open-field/:projet_id/peres/:pere_id', async ({ params }) => {
  await oneOr404('SELECT * FROM "PlantePereOpenField" WHERE id_pere = ? AND id_projet = ?',
    [params.pere_id, params.projet_id], 'Pere non trouve')
  await run('DELETE FROM "PlantePereOpenField" WHERE id_pere = ?', [params.pere_id])
  return { status: 204 }
})
